2단계 - 문자열 덧셈 계산기

heowc opened the pull request:
`2단계 - 문자열 덧셈 계산기`에 대한 작업 결과를 PR 요청드립니다.

heowc committed:
test(StringAddCalculator): StringAddCalculator 클래스에 대한 학습 테스트를 작성

- 기본 뼈대 클래스 추가

heowc committed:
feat(StringAddCalculator): StringAddCalculator 클래스에 대한 기능 구현

- 요구사항에 만족하는 코드 구현

heowc committed:
fix(StringAddCalculator): StringAddCalculator 클래스에 대한 리팩토링

- 메소드 분리
- 들여쓰기 depth 최대 1단계로 줄임
- else 구문 제거

seokhongkim commented on src/main/java/step2/StringAddCalculator.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+class StringAddCalculator {
```
클래스의 접근 제어자와 splitAndSum 메소드의 접근 제어자를 default 로 설정하신 이유가 있을까요?

heowc replied:
타 단계 과제와 연결점이 없다고 판단해서 접근제어자를 `default`로 설정했습니다.

seokhongkim commented on src/main/java/step2/StringAddCalculator.java:
```diff
@@ -0,0 +1,47 @@
+    private static final Pattern DEFAULT_DELIMITER_PATTERN = Pattern.compile("[,:]");
+    private static final Pattern CUSTOM_DELIMITER_PATTERN = Pattern.compile("//(.)\\n(.*)");
+
+    static int splitAndSum(String value) {
```
메소드 이름은 구현보다는 행위를 서술하도록 작성하는 것이 좋습니다.
외부에서 이 메소드를 호출할 때 기대하는 것은 무엇일까요?

heowc replied:
행위를 서술해야한다면 계산을 위한 것이니 `calculate(...)`가 더 적절해보이네요 🤔

seokhongkim commented on src/main/java/step2/StringAddCalculator.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    private static boolean isEmpty(String value) {
+        return value == null || value.isEmpty();
```
꼼꼼한 null check 좋습니다. 👍

seokhongkim commented on src/main/java/step2/StringAddCalculator.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+class StringAddCalculator {
```
한 클래스에 모든 코드가 구현되어 이해하는데 수월한 반면, 모든 변경은 이 클래스를 수정하는 것으로 이뤄질 것 같아요.

코드를 봤을 때, 아래와 같은 역할이 있는 것 같습니다.
1. 문자열을 분할한다. - 일정한 패턴으로 작성되어 있어야 함
2. 각 문자를 숫자 타입으로 변경한다. - 양의 정수여야 함
3. 모든 숫자를 더한다.

작성하신 코드를 객체지향적으로 설계한다면 어떻게 변경해볼 수 있을까요?

heowc replied:
언급해주신 아래 내용들로 클래스를 분리 해볼 수 있을 것 같습니다.

> 1. 문자열을 분할한다. - 일정한 패턴으로 작성되어 있어야 함
> 2. 각 문자를 숫자 타입으로 변경한다. - 양의 정수여야 함
> 3. 모든 숫자를 더한다.

조금 더 자세히 설명해보자면, 

> 1. 문자열을 분할한다. - 일정한 패턴으로 작성되어 있어야 함

기본 패턴과 커스텀 패턴이 있으니 `StringSplitter` 같은 인터페이스를 두고 아래와 같은 클래스를 만들 수 있을 것 같습니다.

- `DefaultStringSplitter`: 기본 패턴으로 자르기
- `CustomStringSplitter`:  커스텀 구분자로 패턴으로 자르기

여기서 좀 더 고민을 해보자면  `StringAddCalculator`에서는 어떻게 분할할지는 관심사가 아니기 때문에 `DelegatingStringSplitter`를 두고 하위에서 적절한 `StringSplitter`를 선택해 자르도록 처리할 수 있을 것 같습니다.

만약 `DelegatingStringSplitter`를 만든다면, `splitter` 패키지를 만들고 관련 인터페이스와 클래스를 하위로 옮겨 `DefaultStringSplitter`와 `CustomStringSplitter`는 default 접근제한자를 활용하여 숨길 수 있을 것 같아요. 

> 2. 각 문자를 숫자 타입으로 변경한다. - 양의 정수여야 함

`StringToPostiveIntCoverter`를 만들어서 분리해볼 수 있을 것 같아요.

> 3. 모든 숫자를 더한다.

`AddOperator`라는 걸 만들어 볼 수 있을 것 같습니다.

seokhongkim replied:
말씀해주신대로 나누면 좋겠네요.
다음 단계를 진행하실 때는 말씀해보신 것처럼 클래스 역할을 나누고, 객체의 협력을 중심으로 구현해보시면 좋겠습니다 😄 

`StringToPositiveIntConverter` 도 좋은 방법이라고 생각하는데, 문자열 덧셈 계산기 맥락에서 `Number` 라는 클래스를 만들어보는 것은 어떨까요?
`Number` 클래스는 문자열 또는 정수를 인자로 받아서 생성하되, 저장하는 값이 음수일 수 없다는 제약조건을 갖는거죠.

다음 단계 미션 하실 때 같이 고민해보시면 좋겠네요!

seokhongkim commented on src/main/java/step2/StringAddCalculator.java:
```diff
@@ -0,0 +1,47 @@
+
+class StringAddCalculator {
+
+    private static final Pattern DEFAULT_DELIMITER_PATTERN = Pattern.compile("[,:]");
```
패턴을 한 번 생성해서 재사용 할 수 있는 것 좋네요 👍

seokhongkim commented on src/main/java/step2/StringAddCalculator.java:
```diff
@@ -0,0 +1,47 @@
+    private static int parseUnsignedInt(String maybeNumber) {
+        int number = Integer.parseInt(maybeNumber);
+        if (number < 0) {
+            throw new RuntimeException("음수가 될 수 없습니다");
```
RuntimeException 보다 상황에 맞는 적절한 예외를 사용하는 게 좋을 것 같아요!

heowc replied:
다른 적절한 예외로 변경하도록하겠습니다 👍

seokhongkim reviewed (changes requested):
원철님, 문자열 덧셈 계산기 잘 구현해주셨습니다.
코멘트 몇 개 남겼으니 확인 부탁드려요!
그리고 요구사항을 todo list 로 작성해야 하는 것 같습니다.

궁금하신 점이나 이야기 나누고 싶은 부분이 있다면 깃헙 코멘트, 슬랙 DM 편하게 남겨주세요!

heowc committed:
fix(StringAddCalculator): 리뷰 내용 반영

- 메소드명 변경: `splitAndSum(...)` -> `calculate(...)`
- `RuntimeException`을 `NumberFormatException` 으로 변경

heowc committed:
docs: 요구사항 문서 작성

seokhongkim commented on README.md:
```diff
@@ -6,4 +6,20 @@
+
+## 요구사항
+
+### 기능 요구사항
```
기능 요구사항을 잘 작성해주셨어요! 👍 
다음에 작성하실 때는 마크다운 문법 중 체크리스트인 `- [ ]` 와 `- [x]`(처리 완료)를 사용해보셔도 좋을 것 같아요.

seokhongkim reviewed (approved):
2단계 미션 진행하느라 고생하셨습니다.

남겨 놓은 피드백은 다음 단계 미션을 진행하실 때 고민해보셔도 될 것 같아서, 이만 PR 머지하도록 하겠습니다.
다음 단계에서 뵙겠습니다!

heowc commented:
@seokhongkim 피드백 감사합니다 :)
